proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,235 @@
+accounts are updated, where `M` corresponds to how many blocks will be in
+the reward distribution phase, which is described in the next section.
+
+
+### Reward Distribution
+
+Reward distribution phase happens right after reward computation phase and
+lasts for `M` blocks.
+
+Based on current validator client's performance, 64 additional accounts can be
+stored per tick time without much performance impact for the overall block
+time. Given that each block time is 64 ticks time long, a total of 4,096
+accounts can be distributed per block. The total number of blocks needed tto
+distributed rewards is
+
+```
+M = ((4096 - 1)+num_stake_accounts)/4096
+```
+
+During the reward distribution phase, rewards will be distributed from the
+`EpochRewardReserve` sysvar account with the address
+`SysvarRewardReserve_{i_th}_111111111` for the `i_th` block. As the
+stake_accounts receive their rewards from the reserve account, the balance of
+the EpochRewardReserve will decrease. Once all rewards have been distributed, the
+balance of the EpochRewardReserve account should be reduced to 0. Before each
+reward distribution, the `EpochRewardReserve` account's balance is check to
+make sure there is enough balance to distribute the reward. After a reward is
+distributed, the balance in `EpochRewardReserve` account is decreased by the
+amount of the reward that are distributed. The hash of reward_account pubkey and reward amount are accumulated when rewards are paying out.
+After all the rewards are
+distributed from the reserve account for the block, the hash computed from all the rewards
+distribution is verified against the hash stored in the `EpochReserve` to make sure
+they match.
+
+
+To ensure that each block distributes a subset of the rewards in a deterministic
+manner for the current epoch, while also randomizing the distribution across
+different epochs, the partitioning of all rewards will be done in a specific
+way. This will involve hashing the account's public key with the epoch number,
+sorting the resulting values, and dividing them into `M` bins. This process will
+ensure that the reward distribution is deterministic for the current epoch,
+while also introducing a degree of randomness between epochs.
+
+### Restrict Stake Account access
+
+To avoid the complexity and potential rewards in unexpected stake accounts, all
+write access to stake accounts are restricted. That means any update to stake
+accounts has to wait till the rewards distribution end. Any transactions
+referencing the native staking program during the epoch rewarding phase will
+fail with a unique transaction error code -
+"LockedRewardAccountsDuringEpochReward".
+
+
+### Snapshot and Cluster Restart
+
+Due to the fact that rewards are distributed over multiple blocks, snapshots taken
+during reward distribution phase for cluster restart, need to store the
+reward calculation result.
+
+To accommodate this, a new field, `Option<RewardsResult>`, is added to the
```

Comment:
I feel the remainder of this section can be omitted from this doc, since it's Labs validator client specific